kernel/tclapi.cc

```diff
@@ -0,0 +1,449 @@
+
+		yosys_tcl_repl_active = in_repl;
+		log_cmd_error_throw = restore_log_cmd_error_throw;
+		return TCL_ERROR;
+	} catch (...) {
+		log_error("uncaught exception during Yosys command invoked from TCL\n");
+	}
+
+	yosys_tcl_repl_active = in_repl;
+	log_cmd_error_throw = restore_log_cmd_error_throw;
+
+	auto &scratchpad = yosys_get_design()->scratchpad;
+	auto result = scratchpad.find("result.json");
+	if (result != scratchpad.end()) {
+		std::string err;
+		auto json = json11::Json::parse(result->second, err);
+		if (err.empty()) {
+			Tcl_SetObjResult(interp, json_to_tcl(interp, json));
+		} else
+			log_warning("Ignoring result.json scratchpad value due to parse error: %s\n", err.c_str());
+	} else if ((result = scratchpad.find("result.string")) != scratchpad.end()) {
+		Tcl_SetObjResult(interp, Tcl_NewStringObj(result->second.data(), result->second.size()));
+	}
+
+	return TCL_OK;
+}
+
+#define FLAG(name) \
+	if (!strcmp(argv[i], "-" #name)) { \
+		name##_flag = true; \
+		continue; \
+	} \
+
+#define ERROR(str) \
+	{ \
+		Tcl_SetResult(interp, (char *)(str), TCL_STATIC); \
+		return TCL_ERROR; \
+	}
+
+static int tcl_get_attr(ClientData, Tcl_Interp *interp, int argc, const char *argv[])
+{
+	int i;
+	bool mod_flag = false, string_flag = false, int_flag = false, bool_flag = false;
+	for (i = 1; i < argc; i++) {
+		FLAG(mod)
+		FLAG(string)
+		FLAG(int)
+		FLAG(bool)
+		break;
+	}
+
+	if ((mod_flag && i != argc - 2) ||
+			(!mod_flag && i != argc - 3) ||
+			(string_flag + int_flag + bool_flag > 1))
+		ERROR("bad usage: expected \"get_attr -mod [-string|-int|-bool] <module> <attrname>\""
+			  " or \"get_attr [-string|-int|-bool] <module> <identifier> <attrname>\"")
+
+	IdString mod_id, obj_id, attr_id;
+	mod_id = RTLIL::escape_id(argv[i++]);
+	if (!mod_flag)
+		obj_id = RTLIL::escape_id(argv[i++]);
+	attr_id = RTLIL::escape_id(argv[i++]);
+
+	RTLIL::Module *mod = yosys_design->module(mod_id);
+	if (!mod)
+		ERROR("module not found")
+
+	RTLIL::AttrObject *obj = nullptr;
+	if (mod_flag) {
+		obj = mod;
+	} else {
+		obj = mod->wire(obj_id);
+		if (!obj)
+			obj = mod->memories.at(obj_id, nullptr);
+		if (!obj)
+			obj = mod->cell(obj_id);
+		if (!obj)
+			obj = mod->processes.at(obj_id, nullptr);
+	}
+
+	if (!obj)
+		ERROR("object not found")
+
+	if (string_flag) {
+		Tcl_SetResult(interp, (char *) obj->get_string_attribute(attr_id).c_str(), TCL_VOLATILE);
+	} else if (int_flag) {
+		if (!obj->has_attribute(attr_id))
+			ERROR("attribute missing (required for -int)");
+
+		RTLIL::Const &value = obj->attributes.at(attr_id);
+		if (value.size() > 32)
+			ERROR("value too large")
+
+		// FIXME: 32'hffffffff will return as negative despite is_signed=false
+		Tcl_SetResult(interp, (char *) std::to_string(value.as_int()).c_str(), TCL_VOLATILE);
+	} else if (bool_flag) {
+		bool value = obj->get_bool_attribute(attr_id);
+		Tcl_SetResult(interp, (char *) std::to_string(value).c_str(), TCL_VOLATILE);
```

Comment:
I think using `Tcl_SetObjResult(interp, Tcl_NewIntObj(value))` here and elsewhere would be nicer and avoid the roundtrip to a string representation.